# Solutions/Objective

# We will import the libraries required to complete our task and get the data

# Imports required libraries
** Import pandas, numpy, matplotlib,and seaborn. Then set %matplotlib inline.

In [10]:
import math
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn import preprocessing
from sklearn.preprocessing import normalize
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
from sklearn.model_selection import train_test_split
from sklearn.linear_model import SGDRegressor
from sklearn import linear_model
import seaborn as sns
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
sns.set_style('darkgrid')
%matplotlib inline
# machine learning

from sklearn.linear_model import SGDClassifier
from sklearn.model_selection import train_test_split
from sklearn import preprocessing
from sklearn.model_selection import train_test_split
from sklearn.metrics import make_scorer, accuracy_score,precision_score
from sklearn.metrics import accuracy_score ,precision_score,recall_score,f1_score
from sklearn.model_selection import train_test_split
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report
from sklearn.preprocessing import MinMaxScaler
from sklearn.linear_model import LinearRegression
import statsmodels.api as sm
from sklearn.metrics import mean_absolute_error, r2_score
from sklearn.preprocessing import MinMaxScaler
from sklearn.linear_model import LinearRegression
import statsmodels.api as sm
from sklearn.metrics import mean_absolute_error, r2_score
from sklearn.feature_selection import chi2
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, confusion_matrix



# Get the Data
We'll work with selecting a candidate in Campus recruitment Prediction data csv file It has S.No, Gender, ssc percentage, hsc percentage, degree, work exp, specialization, mba percentage status , salary
** Read iCampus recruitment Prediction data csv file as a DataFrame called df

In [11]:
df = pd.read_csv("Placement_Data_Full_Class-1.csv")



# Task: Perform exploratory analysis on the data and describe your understanding of the data
. 
Let us quickly view the first few rows and extract some prelimindary information about the dataset


Lets have a look at data using head method

In [3]:
df.head()
print(df)

     sl_no gender  ssc_p    ssc_b  hsc_p    hsc_b     hsc_s  degree_p  \
0        1      M  67.00   Others  91.00   Others  Commerce     58.00   
1        2      M  79.33  Central  78.33   Others   Science     77.48   
2        3      M  65.00  Central  68.00  Central      Arts     64.00   
3        4      M  56.00  Central  52.00  Central   Science     52.00   
4        5      M  85.80  Central  73.60  Central  Commerce     73.30   
..     ...    ...    ...      ...    ...      ...       ...       ...   
210    211      M  80.60   Others  82.00   Others  Commerce     77.60   
211    212      M  58.00   Others  60.00   Others   Science     72.00   
212    213      M  67.00   Others  67.00   Others  Commerce     73.00   
213    214      F  74.00   Others  66.00   Others  Commerce     58.00   
214    215      M  62.00  Central  58.00   Others   Science     53.00   

      degree_t workex  etest_p specialisation  mba_p      status    salary  
0     Sci&Tech     No     55.0         Mkt&HR 

# Lets see the features in given sample data by using columns method

In [4]:
print('='*50)
print("Columns in data")
print('='*50)
print(df.columns.values)

Columns in data
['sl_no' 'gender' 'ssc_p' 'ssc_b' 'hsc_p' 'hsc_b' 'hsc_s' 'degree_p'
 'degree_t' 'workex' 'etest_p' 'specialisation' 'mba_p' 'status' 'salary']


We can see we have below coloums/feature in our dataset
Let's have look at each column information.
sl_no : Serial Number

gender : Candidate gender --> Male='M',Female='F'

ssc_p : SSC (10th) Percentage

ssc_b : SSC Board of Education --> Central (or) Others

hsc_p : HSC (12th) percentage

hsc_b : HSC Board of Education --> Central/ Others

hsc_s : Specialization in HSC

degree_p : Degree Percentage

degree_t : Under Graduation (Degree type)- Field of degree education

workex : Work Experience
etest_p : Employability test percentage ( conducted by college)

specialisation : Post Graduation(MBA)- Specialization

mba_p : MBA percentage

status : Status of placement- Placed/Not placed

salary : Salary offered by corporate to candidates

# Lets find out How many Data we have in our data set using shape method.

In [5]:
print('='*20)
print("Data shape")
print('='*20)
print(df.shape)

Data shape
(215, 15)


We have only 215 samples in our dataset

# Generate descriptive statistics.

In [6]:

# Checking Min, Max, Mean, Std for all the columns
df.describe(percentiles=[.10, .25, .5, .75])

sl_no       ssc_p       hsc_p    degree_p     etest_p       mba_p  \
count  215.000000  215.000000  215.000000  215.000000  215.000000  215.000000   
mean   108.000000   67.303395   66.333163   66.370186   72.100558   62.278186   
std     62.209324   10.827205   10.897509    7.358743   13.275956    5.833385   
min      1.000000   40.890000   37.000000   50.000000   50.000000   51.210000   
10%     22.400000   52.000000   52.000000   57.000000   55.802000   54.986000   
25%     54.500000   60.600000   60.900000   61.000000   60.000000   57.945000   
50%    108.000000   67.000000   65.000000   66.000000   71.000000   62.000000   
75%    161.500000   75.700000   73.000000   72.000000   83.500000   66.255000   
max    215.000000   89.400000   97.700000   91.000000   98.000000   77.890000   

              salary  
count     148.000000  
mean   288655.405405  
std     93457.452420  
min    200000.000000  
10%    217400.000000  
25%    240000.000000  
50%    265000.000000  
75%    300000.000000  
max    940000.000000

# Let see data object Type

In [7]:
print('='*50)
print("\nData Information\n")
print('='*50)
df.info()


Data Information

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 215 entries, 0 to 214
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   sl_no           215 non-null    int64  
 1   gender          215 non-null    object 
 2   ssc_p           215 non-null    float64
 3   ssc_b           215 non-null    object 
 4   hsc_p           215 non-null    float64
 5   hsc_b           215 non-null    object 
 6   hsc_s           215 non-null    object 
 7   degree_p        215 non-null    float64
 8   degree_t        215 non-null    object 
 9   workex          215 non-null    object 
 10  etest_p         215 non-null    float64
 11  specialisation  215 non-null    object 
 12  mba_p           215 non-null    float64
 13  status          215 non-null    object 
 14  salary          148 non-null    float64
dtypes: float64(6), int64(1), object(8)
memory usage: 25.3+ KB


# We have 7 columns with real values and 8 with object datatype

#As it is clear that we don't need sl_no in training model or in EDA.
#Thus I am dropping sl_no column. Rest of them I will keep as it is.
#After performing EDA I will drop other if needed.

In [8]:
data = df.drop(['sl_no'], axis=1)

# Getting the object columns

In [9]:
# column=df.select_dtypes(include=['object'])
for col in column:
    display(data[col].value_counts())

NameError: name 'column' is not defined

# OBSERVATION:

OBSERVATION:
Looks like except for hsc_s and degree_t with 3 classes, all other have 2 classes each and also we can see that this data is slightly imbalanced as we have 148 placed students and 67 not placed students.

# Data Cleaning

# Checking for null Values

In [ ]:
#Check the null values
df.isnull().sum()

# # It is clear that only salary has null columns. Let's see how much?We need to clean the nan value

In [ ]:
missing_percantage=df['salary'].isnull().sum()/len(df)*100

print(round(missing_percantage,2),'%')

This means that around 31% candidates were not placed 

# From above it seems 31.16 % of candidates were not placed.
For perdicting the salary as per our Assignment objective i will hanadle it as below

In [ ]:
df['salary'].fillna(0, inplace=True)



# After cleaning the null value. Lets once again check the null value

In [ ]:

df.isnull().sum()

All null valaues are cleared

# Let Check for Duplicates

In [ ]:
print("Number of duplicate values in the dataset is : ", df.duplicated().sum())

Since duplicate dataset is 0 , No need to handle it.we can proceed with exploring the feature.

# Let Check for Outliers

Outliers are unusual values in your dataset, and they can distort statistical analyses and violate their assumptions.
let's first visualize our data and decide on what to do with the outliers

In [ ]:
plt.figure(figsize = (15, 10))
plt.style.use('seaborn-white')
ax=plt.subplot(221)
plt.boxplot(df['ssc_p'])
ax.set_title('Secondary school percentage')
ax=plt.subplot(222)
plt.boxplot(df['hsc_p'])
ax.set_title('Higher Secondary school percentage')
ax=plt.subplot(223)
plt.boxplot(df['degree_p'])
ax.set_title('UG Degree percentage')
ax=plt.subplot(224)
plt.boxplot(df['etest_p'])
ax.set_title('Employability percentage')

As you see, we have very less number of outliers in our features. Especially we have majority of the outliers in hsc percentage Let's clear em up

In [ ]:
Q1 = df['hsc_p'].quantile(0.25)
Q3 = df['hsc_p'].quantile(0.75)
IQR = Q3 - Q1    #IQR is interquartile range. 

filter = (df['hsc_p'] >= Q1 - 1.5 * IQR) & (df['hsc_p'] <= Q3 + 1.5 *IQR)
placement_filtered=df.loc[filter]

After removing Outlier Graphs look like

In [ ]:
plt.figure(figsize = (15, 5))
plt.style.use('seaborn-white')
ax=plt.subplot(121)
plt.boxplot(df['hsc_p'])
ax.set_title('Before removing outliers(hsc_p)')
ax=plt.subplot(122)
plt.boxplot(placement_filtered['hsc_p'])
ax.set_title('After removing outliers(hsc_p)')

# Perform Exploratory Data Analysis(EDA) & Visualizations of Data by each Features

# Feature:- Status

In [ ]:
plt.figure(figsize = (15,5))
plt.subplot(1,2,1)
df['status'].value_counts().plot(kind = 'pie', autopct = '%.2f%%')
plt.subplot(1,2,2)
sns.countplot(x = 'status', data = df)
plt.show()

# From above plot we can see 68.84 % of students are placed and 31.16 % Student are not placed.

# Feature:-Gender



Let's check whethere gender affect on placement?

In [ ]:
df['gender'].value_counts()

# Out of 215 candidates, 139 are male and 76 are female.

In [ ]:

sns.countplot(x = 'gender', hue='status',data = df)
plt.show()



In [ ]:
pd.DataFrame(df.groupby(['gender','status'])['status'].count())

In [ ]:
plt.figure(figsize=(10,7))
sns.boxplot(y='gender',x='salary',data=df)

# OBSERVATIONS

# Male have high chances of getting placed compared to females.
#We have samples of 139 Male studets and 76 Female students.
#The number of male students are almost double as compared to female.
#More outliers on Male,Male students are getting high CTC jobs.
#Male students are offered slightly greater salary than female on an average

# Feature : ssc_b,hsc_b,hsc_s

# Does the board of education affect placements?

# SSC Board

In [ ]:
df['ssc_b'].value_counts()

In [ ]:
ssc_b_count=pd.DataFrame(data.groupby(['ssc_b','status'])['status'].count())
ssc_b_count

In [ ]:
sns.countplot(x='ssc_b', hue='status', data=df)

# From the above analysis I can say that, SSC board is not important to recruiters when it come to hiring candidates. So I am not going to use this feature while training model.

# HSC Board

In [ ]:
df['hsc_b'].value_counts()

In [ ]:
hsc_board_count = pd.DataFrame(data.groupby(['hsc_b','status'])['status'].count())
hsc_board_count

In [ ]:
sns.countplot(x='hsc_b', hue='status', data=df)

# From the above analysis I can say that, SSC board is not important to recruiters when it come to hiring candidates. So I am not going to use this feature while training model.

# HSC_S Specialisation

In [ ]:
df['hsc_s'].value_counts()

In [ ]:
hsc_specialization_count = pd.DataFrame(data.groupby(['hsc_s','status'])['status'].count())
hsc_specialization_count

In [ ]:
sns.countplot(x='hsc_s', hue='status', data=df)

In [ ]:
plt.figure(figsize=(15,8))
sns.catplot(x='hsc_b',hue='hsc_s',col='status',data=df,kind='count')
plt.show()

# Observetion:for various boards

# Feature : ssc_p,hsc_p,degree_p,mba_p

# Does your academic score influence your chance of placement?

#  SSC Percentage

In [ ]:
sns.distplot(data['ssc_p'], kde=False)
plt.title('Distribution of SSC Percentage')
plt.xlabel('SSC %')


sns.distplot(data.ssc_p[data.status=='Not Placed'])
sns.distplot(data.ssc_p[data.status=='Placed'])
plt.legend(['Not placed','placed'])
plt.show()

In [ ]:
sns.catplot(y='ssc_p', x='status', data=df)
plt.xlabel('Employment Status')
plt.ylabel('SSC %')

Students who are place have higer percentage in SSC

#  HSC Percentage

In [ ]:
sns.distplot(df['hsc_p'], kde=False)
plt.title('Distribution of SSC Percentage')
plt.xlabel('HSC %')

In [ ]:
sns.catplot(y='hsc_p', x='status', data=df)
plt.xlabel('Employment Status')
plt.ylabel('HSC %')

HSC percentage are important features. As all placed students have higher percentages.

# Degree Percentage

In [ ]:
sns.distplot(df['degree_p'], kde=False)
plt.title('Distribution of Degree Percentage')
plt.xlabel('Degree %')

In [ ]:
sns.catplot(y='degree_p', x='status', data=df)
plt.xlabel('Employment Status')
plt.ylabel('Degree %')

Like SSC and HSC percentages, Degree Percentages are also impotant factor to get placed.

In [ ]:
plt.figure(figsize = (15, 15))
ax=plt.subplot(221)
sns.boxplot(x='status',y='ssc_p',data=df)
ax.set_title('Secondary school percentage')
ax=plt.subplot(222)
sns.boxplot(x='status',y='hsc_p',data=df)
ax.set_title('Higher Secondary school percentage')
ax=plt.subplot(223)
sns.boxplot(x='status',y='degree_p',data=df)
ax.set_title('UG Degree percentage')
ax=plt.subplot(224)
sns.boxplot(x='status',y='mba_p',data=df)
ax.set_title('MBA percentage')

#Distribution of all percentages
plt.figure(figsize = (15, 7))
plt.style.use('seaborn-white')
plt.subplot(231)
sns.distplot(df['ssc_p'])
fig = plt.gcf()
fig.set_size_inches(10,10)

plt.subplot(232)
sns.distplot(df['hsc_p'])
fig = plt.gcf()
fig.set_size_inches(10,10)

plt.subplot(233)
sns.distplot(df['degree_p'])
fig = plt.gcf()
fig.set_size_inches(10,10)



plt.subplot(235)
sns.distplot(df['mba_p'])
fig = plt.gcf()
fig.set_size_inches(10,10)

#plt.subplot(236)
#sns.distplot(df['salary'])
#fig = plt.gcf()
#fig.set_size_inches(10,10)

# OBSERVATION For Percentage :-

# Degree Type



In [ ]:
df['degree_t'].value_counts()

In [ ]:
degree_count = pd.DataFrame(data.groupby(['degree_t','status'])['status'].count())
degree_count

In [ ]:
sns.countplot(x='degree_t', hue='status', data=df)

# Work Experience

In [ ]:
df['workex'].value_counts()

In [ ]:
work_exp= pd.DataFrame(data.groupby(['workex','status'])['status'].count())
work_exp

In [ ]:
sns.countplot(x='workex', hue='status', data=df)

# OBSERVATION From Work Experience

# It is clear that candidate with work experience have higher chance of getting placed.

# Employment Test Percentage

In [ ]:
sns.distplot(df['etest_p'], kde=False)
plt.title('Distribution of MBA Percentage')
plt.xlabel('Employment Test %')

In [ ]:
sns.catplot(y='etest_p', x='status', data=df)
plt.xlabel('Employment Status')
plt.ylabel('Employment Test %')

# MBA Specialization

In [ ]:
df['specialisation'].value_counts()

In [ ]:
specialization_count_placed= pd.DataFrame(data.groupby(['specialisation','status'])['status'].count())
specialization_count_placed

In [ ]:
sns.countplot(x='specialisation', hue='status', data=df)
plt.xlabel('MBA Specialization')

# From above we can see student who has opted MBA in Mkt and Finance has higher chnance of placed and get salary

# MBA Percentage

In [ ]:
sns.distplot(df['mba_p'], kde=False)
plt.title('Distribution of MBA Percentage')
plt.xlabel('MBA %')

In [ ]:
sns.catplot(y='mba_p', x='status', data=df)
plt.xlabel('Employment Status')
plt.ylabel('MBA %')

# Most Important Factor; Salary

In [ ]:
df['salary'].fillna(-1, inplace=True)
sns.distplot(df['salary'], kde=False)
plt.title('Distribution of Salary')
plt.xlabel('Salary')

plt.figure(figsize=(10,5))
sns.distplot(df['salary'], bins=50, hist=False)
plt.title("Salary Distribution")
plt.show()

In [ ]:
sns.catplot(y='salary', x='status', data=df)
plt.xlabel('Employment Status')
plt.ylabel('Salary')

# Observation from salary :-

# Looking at the distribution we can say that the most of the students get a package between 200k-400k and most salaries above 400,000 are outliers.


Male candidates are making more money as compared to female candidates.
Only one candidate got around 10L PA
The average of the salary is a little more than 2LPA

#  Correlation between features & Various Plots

 Correlation between features

In [ ]:
sns.pairplot(data=df[['ssc_p','hsc_p','degree_p', 'etest_p','mba_p','salary', 'status']], hue="status", diag_kind='hist')

# Now we will convert all string data types into numeric 

In [ ]:
# copy original dataset into another varialble to build model without feature engineering
df_without_fe = df.copy(deep=True)

In [ ]:
df_without_fe.dtypes

# Convert string data into categoary data type

In [ ]:
df_without_fe["gender"] = df_without_fe["gender"].astype('category')
df_without_fe["ssc_b"] = df_without_fe["ssc_b"].astype('category')
df_without_fe["hsc_b"] = df_without_fe["hsc_b"].astype('category')
df_without_fe["hsc_s"] = df_without_fe["hsc_s"].astype('category')
df_without_fe["degree_t"] = df_without_fe["degree_t"].astype('category')
df_without_fe["workex"] = df_without_fe["workex"].astype('category')
df_without_fe["specialisation"] = df_without_fe["specialisation"].astype('category')
df_without_fe["status"] = df_without_fe["status"].astype('category')

In [ ]:
df_without_fe.dtypes

#  Convert  category data type to numeric data type

In [ ]:
# Create dictionary of each category
dict_gender = dict(enumerate(df_without_fe["gender"].cat.categories))
dict_ssc_b = dict(enumerate(df_without_fe["ssc_b"].cat.categories))
dict_hsc_b = dict(enumerate(df_without_fe["hsc_b"].cat.categories))
dict_hsc_s = dict(enumerate(df_without_fe["hsc_s"].cat.categories))
dict_degree_t = dict(enumerate(df_without_fe["degree_t"].cat.categories))
dict_workex = dict(enumerate(df_without_fe["workex"].cat.categories))
dict_specialisation = dict(enumerate(df_without_fe["specialisation"].cat.categories))
dict_status = dict(enumerate(df_without_fe["status"].cat.categories))

In [ ]:
print('Dictionary of gender          : ', dict_gender)
print('Dictionary of ssc_b           : ', dict_ssc_b)
print('Dictionary of hsc_b           : ', dict_hsc_b)
print('Dictionary of hsc_s           : ', dict_hsc_s)
print('Dictionary of degree_t        : ', dict_degree_t)
print('Dictionary of workex          : ', dict_workex)
print('Dictionary of specialisation  : ', dict_specialisation)
print('Dictionary of status          : ', dict_status)

In [ ]:
df_without_fe["gender"] = df_without_fe["gender"].cat.codes
df_without_fe["ssc_b"] = df_without_fe["ssc_b"].cat.codes
df_without_fe["hsc_b"] = df_without_fe["hsc_b"].cat.codes
df_without_fe["hsc_s"] = df_without_fe["hsc_s"].cat.codes
df_without_fe["degree_t"] = df_without_fe["degree_t"].cat.codes
df_without_fe["workex"] = df_without_fe["workex"].cat.codes
df_without_fe["specialisation"] = df_without_fe["specialisation"].cat.codes
df_without_fe["status"] = df_without_fe["status"].cat.codes

In [ ]:
df_without_fe.dtypes

In [ ]:
df_without_fe

# Linear regression-Prediction of Salary Before Feature Engineering


#Spliting data into train and test

In [ ]:
feature_columns = ['gender', 'ssc_p', 'ssc_b', 'hsc_p', 'hsc_b', 'hsc_s', 'degree_p', 'degree_t', 'workex', 'etest_p', 'specialisation', 'mba_p']
output_column = ['salary']

X = df_without_fe[feature_columns]
y = df_without_fe[output_column]

#y = df_without_fe['salary']
#X = df_without_fe.copy().drop('status', axis=1)
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=101)

**As different columns have different range of values hence dominance of few columns will be there in dataset. To avoid this we are converting it into standarize format**

In [ ]:

from sklearn import preprocessing
scaler = preprocessing.StandardScaler().fit(X_train)
X_train = scaler.transform(X_train)
X_test=scaler.transform(X_test)
X_train

Train model using linear regression

In [ ]:
regr = linear_model.LinearRegression()
#regr.fit(X_train, Y_train.ravel())
regr.fit(X_train,y_train)

print('Coefficients: \n', regr.coef_)
print('Intercept: \n', regr.intercept_)



# Predicting Test Data & Execute ML model on testing data

In [ ]:
predictions = regr.predict( X_test)
predictions


In [ ]:
plt.scatter(y_test,predictions)
plt.xlabel('y_test')
plt.ylabel('Predicted Y')


# Evaluating the Model

In [ ]:
# calculate these metrics by hand!
from sklearn import metrics

print('MAE:', metrics.mean_absolute_error(y_test, predictions))
print('MSE:', metrics.mean_squared_error(y_test, predictions))
print('RMSE:', np.sqrt(metrics.mean_squared_error(y_test, predictions)))
print('R2:',metrics.explained_variance_score(y_test, predictions))

print('Accuracy Before FE:',regr.score(X_test,y_test))

Accuray of model is approx 30.11% 

# We will also try linear regression,With OLS

In [ ]:
feature_columns = ['gender', 'ssc_p', 'ssc_b', 'hsc_p', 'hsc_b', 'hsc_s', 'degree_p', 'degree_t', 'workex', 'etest_p', 'specialisation', 'mba_p']
output_column = ['salary']

x= df_without_fe[feature_columns]
y = df_without_fe[output_column]
#y = df_without_fe['salary']
#x = df_without_fe.copy().drop('status', axis=1)
x.head()

In [ ]:
train_x, test_x, train_y, test_y = train_test_split(x,y , test_size=0.2, random_state=42)

print(train_x.shape)
print(test_x.shape)

In [ ]:
linear_model = sm.OLS(train_y,train_x.astype(float))
results = linear_model.fit()
results.params

In [ ]:
print(results.summary())

In [ ]:
pred_y = results.predict(test_x)
# print(pred_y[:10])
# print(test_y[:10])

col = ['actual','prediction']

prediction = pd.concat([test_y,pred_y],axis=1)
prediction.columns = col
prediction

In [ ]:
_, ax = plt.subplots()

ax.scatter(x = range(0, test_y.size), y=test_y, c = 'blue', label = 'Actual', alpha = 0.3)
ax.scatter(x = range(0, pred_y.size), y=pred_y, c = 'red', label = 'Predicted', alpha = 0.3)

plt.title('Actual and predicted values')
plt.xlabel('Observations')
plt.ylabel('Salary')
plt.legend()
plt.show()

# Metrics With OLS withOut Feature Engineering

In [ ]:
print('MAE:', metrics.mean_absolute_error(test_y, pred_y))
print('MSE:', metrics.mean_squared_error(test_y, pred_y))
print('RMSE:', np.sqrt(metrics.mean_squared_error(test_y, pred_y)))
print('R2:',metrics.explained_variance_score(test_y, pred_y))



# Feature Engineering

# Feature mapping and generation

Let's drop all unwanted columns as menstioned in above section.

SSC Board
HSC Board
HSC Specialisation
Degree Type
Salart

In [ ]:
df.drop(['ssc_b','hsc_b', 'hsc_s', 'degree_t'], axis=1, inplace=True)

# Let's map categorical feature to numeric one. Categorical features:

# Label encoding

In [ ]:
from sklearn.preprocessing import LabelEncoder

# exrtact categorical features

cat_col=df.select_dtypes(object).columns.tolist()
len(cat_col)

In [ ]:
# Convert categorized values to numerical values

le = LabelEncoder()
df[cat_col] =df[cat_col].astype('str').apply(le.fit_transform)
df

In [ ]:
df["gender"] = data.gender.map({"M":0,"F":1})
df["workex"] = data.workex.map({"No":0, "Yes":1})
df["status"] = data.status.map({"Not Placed":0, "Placed":1})
df["specialisation"] = data.specialisation.map({"Mkt&HR":0, "Mkt&Fin":1})

In [ ]:
df.columns

In [ ]:
df.head()
print(df)

# 3: Apply 2 feature selection engineering techniques and Compare the two selected feature engineering techniques. 

In [ ]:
 #Chi-Sqaured for feature selection

In [ ]:
chi2_scores, p_values = chi2(df.iloc[:,:-2],df.iloc[:,-2])
print("Columns of dataframe: \b", df.iloc[:,:-2].columns)
print("Chi-Squared Scores : \n",chi2_scores)
print("\np values : \n",p_values)

In [ ]:
print("{:25} {:25} {:25}".format('FeatureName','Chi-Squared Scores', 'p values'))
for f_name, chi_score, p_value in zip(df.iloc[:,:-2], chi2_scores, p_values):
    print("{:25} {:25} {:25}".format(str(f_name), str(chi_score), str(p_value)))

- If value of Chi-Squared score is higher it means output is highly dependent on that feature. Please find key updates regarding feature
- **ssc_p :** Chi-Squared score is the highest for this features that is **137.73925800375304**. Becuse of that this is the most important feature
- **hsc_p :** This is the second most important feature for output with Chi-Squared score of **92.44931180755225**
- **mba_p :** This is least important feature as per the Chi-Squared score which **0.6918567933507178**

- **Please find below importance of feature from higher to lower using Chi-Squared score**

| Feature name | Chi-Squared score |
| --- | -: |
|	gender                    	|	5.532155289618014         	|
|	ssc_p                     	|	137.73925800375304                    	|
|	hsc_p                                          	|	92.44931180755225	|
|	degree_p	|	40.20489567530576|
|	workex                    	|	10.745483684572026	|
|	etest_p                   	|	8.522678737307453	|
|	specialisation            	|	5.968657556810542	|
|	mba_p                     	|	0.6918567933507178	|




# Correlation Coefficient

Correlation is a statistical technique that can show whether and how strongly pairs of variables are related

In [ ]:
plt.figure(figsize=(12,8))
sns.heatmap(df.corr(),annot=True)
plt.show()

Candidates with Higher MBA Percentage have a higher chance of getting Placed.

Again Higher SSC Percentage is also beneficial for Placement.

Higher HSC Percentage and Degree Percentage also plays a Significant role in Placement.

Surprisingly E Test Percentage does not affect a candidate's Placement Outcome.

Comparing feture selection techniques : Chi-Sqaured for feature selection vs correlation coefficient 
- **Chi-Sqaured for feature selection :** This technique provided five best features those are `ssc_p                     , hsc_p , workex , degree_p, etest_p `
- **Correlation Coefficient :** This technique provided five best features those are `ssc_p, hsc_p, hsc_s, degree_p, workex`


# 4.1. Additional feature generation

In [ ]:
def new_features(df):
    df['hsc_to_ssc'] = df['hsc_p'] / df['ssc_p']
    df['degree_to_hsc'] = df['degree_p'] / df['hsc_p']
    df['degree_to_ssc'] = df['degree_p'] / df['ssc_p']
    df['mba_to_degree'] = df['mba_p'] / df['degree_p']
    df['mba_to_etest'] = df['mba_p'] / df['etest_p']
        
    return df

In [ ]:
fe_data = new_features(df)

# Feature selection


We will now select the features (X) for our model. These features will help our model identify patterns. The features will be columns.

"When feature engineering is done, we usually tend to decrease the dimensionality by selecting the "right" number of features that capture the essential."

# Linear regression-Prediction of Salary After Feature Engineering


# Train and Test Split 

In [ ]:
feature_columns = [ 'ssc_p', 'hsc_p', 'degree_p',  'workex', 'etest_p', 'specialisation', ]
output_column = ['salary']

X= fe_data[feature_columns]
y = fe_data[output_column]

# Seperating Features and Target
#df_with_fe = df.copy(deep=True)
#X = df_with_fe.copy().drop('status', axis=1)
#y = df_with_fe['salary']


In [ ]:
#Train Test Split
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=101)

**As different columns have different range of values hence dominance of few columns will be there in dataset. To avoid this we are converting it into standarize format**

In [ ]:

from sklearn import preprocessing
scaler = preprocessing.StandardScaler().fit(X_train)
X_train = scaler.transform(X_train)
X_test=scaler.transform(X_test)
X_train

# Now we will train several Machine Learning models and compare their results. 

# Linear regression-Prediction of Salary After Preprocessing

In [ ]:
from sklearn.linear_model import LinearRegression
lm = LinearRegression()
lm.fit(X_train,y_train)

print('Coefficients: \n', lm.coef_)
print('Intercept: \n', lm.intercept_)



# Predicting Test Data

In [ ]:
predictions = lm.predict( X_test)

# ** Create a scatterplot of the real test values versus the predicted values. **

In [ ]:
plt.scatter(y_test,predictions)
plt.xlabel('y_test')
plt.ylabel('Predicted Y')


# Evaluating the Model

Let's evaluate our model performance by calculating the residual sum of squares and the explained variance score (R^2).

** Calculate the Mean Absolute Error, Mean Squared Error, and the Root Mean Squared Error. Refer to the lecture or to Wikipedia for the formulas**

In [ ]:
# calculate these metrics by hand!
from sklearn import metrics

print('MAE:', metrics.mean_absolute_error(y_test, predictions))
print('MSE:', metrics.mean_squared_error(y_test, predictions))
print('RMSE:', np.sqrt(metrics.mean_squared_error(y_test, predictions)))
print('R2:',metrics.explained_variance_score(y_test, predictions))
#print('Accuracy After FE:',regr.score(X_test,y_test))
print('Accuracy After FE:',lm.score(X_test,y_test))

# Model With OLS approch After Feature Engineering

In [ ]:
# Seperating Features and Target
feature_columns = [ 'ssc_p', 'hsc_p', 'degree_p',  'workex', 'etest_p', 'specialisation', ]
output_column = ['salary']

x= fe_data[feature_columns]
y = fe_data[output_column]


#df_with_fe = df.copy(deep=True)
#x = df_with_fe.copy().drop('status', axis=1)
#y = df_with_fe['salary']

In [ ]:
train_x, test_x, train_y, test_y = train_test_split(x,y , test_size=0.2, random_state=42)

print(train_x.shape)
print(test_x.shape)

In [ ]:
linear_model = sm.OLS(train_y,train_x.astype(float))
results = linear_model.fit()
results.params

In [ ]:
print(results.summary())

In [ ]:
pred_y = results.predict(test_x)
# print(pred_y[:10])
# print(test_y[:10])

col = ['actual','prediction']

prediction = pd.concat([test_y,pred_y],axis=1)
prediction.columns = col
prediction

In [ ]:
_, ax = plt.subplots()

ax.scatter(x = range(0, test_y.size), y=test_y, c = 'blue', label = 'Actual', alpha = 0.3)
ax.scatter(x = range(0, pred_y.size), y=pred_y, c = 'red', label = 'Predicted', alpha = 0.3)

plt.title('Actual and predicted values')
plt.xlabel('Observations')
plt.ylabel('Salary')
plt.legend()
plt.show()

# Metircs with OLS and Feature Engineering

In [ ]:
print('MAE:', metrics.mean_absolute_error(test_y, pred_y))
print('MSE:', metrics.mean_squared_error(test_y, pred_y))
print('RMSE:', np.sqrt(metrics.mean_squared_error(test_y, pred_y)))
print('R2:',metrics.explained_variance_score(test_y, pred_y))


# Best Performing Model?

# **Details of Linear Regression Before Feature Engineering**

 MAE: 103097.5828893746,MSE: 23193427141.787605 ,RMSE: 152293.88412470018, R2: 0.31554326001707655 
# Accuracy Before FE: 0.3011999997682673

#**Details of Linear Regression OLS Approch Before Feature Engineering**

MAE: 93772.9896712033,MSE: 14795559262.604927,RMSE: 121636.99791841678,R2: 0.330547092279012

# **Details of Linear Regression After Feature Engineering**

MAE: 102429.04029170859,MSE: 22435751119.35446,RMSE: 149785.6839599648,R2: 0.33660707882277774
# Accuracy After FE: 0.324028191626896

#**Details of Linear Regression OLS Approch Before Feature Engineering**


MAE: 4.114506719828394e-10,MSE: 1.8157462943665248e-19,RMSE: 4.261157465251108e-10,R2: 0.27202667648457324

# From Above we can see the benefit of Feature Engineering, Data with Feature engineering giving better performance and accuracy. However it seems data was already in good shape because there is no much difference in performance.

# Answers of objective

### 5.1: Which factor influenced a candidate in getting placed?
- As per Chi-Squared score **ssc_p** is the most influence factor getting placed
- Additionally, **hsc_p, degree_p, workex** feature are from top  features by Chi-Squared and corelation coeficient 


### 5.2: Do percentage matters for one to get placed? 
- Yes, percentage is matters while getting placed. ` ssc_p, degree_p, hsc_p` this three part of top five features.
- Additionally, as per Chi-Squared score `mba_p` is matter less and this is second least important feature from all features

### 5.3: Which degree specialization is much demanded by corporate? 




# We have seen in while exploring each feature  Mkt&Fin is the more demanded by corporate

# Conclusion